File: plugins/bubblellaneous/registry/furniture.py
from plugins.bubblellaneous.internal import (Block, BlockData, BlockMaterials,
                                             BlockType)
from plugins.bubblellaneous.internal.unit.variant import BlockVariant


class BubbleBench(Block):
    """
    :yellow [☶ Description]
    Used for obtaining all custom blocks & items.

    :green [☄ Instructions]
    Open the block, use compass inside to view all help tooltips.
    """

    base = BlockData.Base.CONTAINER
    sound = BlockData.Sound.INDUSTRIAL
    facing = BlockData.Facing.PLAYER
    recipe = [
        BlockData.RecipeEntry("block", "crafting_table"),
        BlockData.RecipeEntry("item", "glass_bottle"),
    ]
    tags = [
        BlockData.Uses.GUI,
        BlockData.Uses.BRIGHTNESS_FIX,
    ]


class Table(BlockVariant):
    """
    :yellow [☶ Description]
    Empty.

    :light_purple [⇄ Connectivity]
    Can connect to other tables.
    """

    material = BlockMaterials.SOLID

    base = BlockData.Base.UPPER
    sound = BlockData.Sound.WOOD
    facing = BlockData.Facing.NONE
    recipe = [
        BlockData.RecipeEntry("block", "[base]", 1),
    ]
    tags = [
        BlockData.Uses.PLACE,
        BlockData.Uses.BLOCKSTATES,
    ]
    blockstates = BlockData.BlockStates(
        "@self",
        BlockData.State("default"),
        BlockData.State("side", "0100~~:0", "1000~~:180", "0010~~:-90", "0001~~:90"),
        BlockData.State("corner", "1010~~:0", "1001~~:-90", "0110~~:90", "0101~~:180"),
        BlockData.State("middle", "11~~~~", "~~11~~"),
    )


class Shelf(BlockVariant):
    """
    :yellow [☶ Description]
    Can store a single item stack.
    """

    material = BlockMaterials.SOLID
    block_type = BlockType.shelf(amount=1)

    base = BlockData.Base.VOID
    sound = BlockData.Sound.WOOD
    facing = BlockData.Facing.WALL_NORMAL
    recipe = [
        BlockData.RecipeEntry("block", "[base]", 1),
    ]
    tags = [
        BlockData.Uses.PLACE,
    ]


class Cabinet(BlockVariant):
    """
    :yellow [☶ Description]
    A regular container (27 slots).
    """

    material = BlockMaterials.SOLID

    base = BlockData.Base.CONTAINER
    sound = BlockData.Sound.WOOD
    facing = BlockData.Facing.PLAYER
    recipe = [
        BlockData.RecipeEntry("block", "[base]", 1),
    ]
    tags = [
        BlockData.Uses.BRIGHTNESS_FIX,
    ]


class Drawer(BlockVariant):
    """
    :yellow [☶ Description]
    Empty.
    """

    material = BlockMaterials.SOLID

    base = BlockData.Base.UPPER
    sound = BlockData.Sound.WOOD
    facing = BlockData.Facing.PLAYER
    recipe = [
        BlockData.RecipeEntry("block", "[base]", 1),
    ]
    tags = []


class Cupboard(BlockVariant):
    """
    :yellow [☶ Description]
    Can store up to 2 item stacks.
    """

    material = BlockMaterials.SOLID
    block_type = BlockType.shelf(amount=2)

    base = BlockData.Base.SOLID
    sound = BlockData.Sound.WOOD
    facing = BlockData.Facing.PLAYER
    recipe = [
        BlockData.RecipeEntry("block", "[base]", 1),
    ]
    tags = [
        BlockData.Uses.PLACE,
        BlockData.Uses.BRIGHTNESS_FIX,
    ]


class Bench(BlockVariant):
    """
    :yellow [☶ Description]
    A seat.
    """

    material = BlockMaterials.SOLID
    block_type = BlockType.seat(width=0.8, height=0.6)

    base = BlockData.Base.TRIPWIRE
    sound = BlockData.Sound.WOOD
    facing = BlockData.Facing.PLAYER
    recipe = [
        BlockData.RecipeEntry("block", "[base]", 1),
    ]
    tags = [
        BlockData.Uses.PLACE,
    ]


class Couch(BlockVariant):
    """
    :yellow [☶ Description]
    A seat.

    :light_purple [⇄ Connectivity]
    Can connect to other couches with corners support.
    """

    material = BlockMaterials.WOOL
    block_type = BlockType.seat(width=0.8, height=0.5)

    base = BlockData.Base.TRIPWIRE
    sound = BlockData.Sound.WOOL
    facing = BlockData.Facing.PLAYER
    recipe = [
        BlockData.RecipeEntry("block", "[color]", 6),
    ]
    tags = [
        BlockData.Uses.PLACE,
        BlockData.Uses.BLOCKSTATES,
    ]
    blockstates = BlockData.BlockStates(
        "@self",
        BlockData.State("default"),
        BlockData.State("left", "0010~~"),
        BlockData.State("right", "0001~~"),
        BlockData.State("middle", "0011~~"),
        BlockData.State("corner", "1010~~:90", "1001~~:0"),
        BlockData.State("slab", "~1~~~~"),
    )


class Stool(BlockVariant):
    """
    :yellow [☶ Description]
    A seat.
    """

    material = BlockMaterials.WOOD
    block_type = BlockType.seat(width=0.8, height=0.6)

    base = BlockData.Base.VOID
    sound = BlockData.Sound.WOOD
    facing = BlockData.Facing.NONE
    recipe = [
        BlockData.RecipeEntry("block", "[base]", 1),
    ]
    tags = [
        BlockData.Uses.PLACE,
    ]


class Barstool(BlockVariant):
    """
    :yellow [☶ Description]
    A seat.
    """

    material = BlockMaterials.WOOD
    block_type = BlockType.seat(width=0.8, height=0.7)

    base = BlockData.Base.VOID
    sound = BlockData.Sound.WOOD
    facing = BlockData.Facing.NONE
    recipe = [
        BlockData.RecipeEntry("block", "[base]", 1),
    ]
    tags = [
        BlockData.Uses.PLACE,
    ]


class DinningChair(BlockVariant):
    """
    :yellow [☶ Description]
    A seat.
    """

    material = BlockMaterials.WOOD_WITH_WOOL
    block_type = BlockType.seat(width=0.8, height=0.4)

    base = BlockData.Base.VOID
    sound = BlockData.Sound.WOOD
    facing = BlockData.Facing.PLAYER
    recipe = [
        BlockData.RecipeEntry("block", "[base]", 1),
        BlockData.RecipeEntry("block", "[color]", 2),
    ]
    tags = [
        BlockData.Uses.PLACE,
    ]


class Blinds(BlockVariant):
    """
    :yellow [☶ Description]
    Empty.
    """

    material = BlockMaterials.SOLID

    base = BlockData.Base.VOID
    sound = BlockData.Sound.WOOD
    facing = BlockData.Facing.WALL_NORMAL
    recipe = [
        BlockData.RecipeEntry("block", "[base]", 1),
    ]
    tags = []


class KitchenCabinet(BlockVariant):
    """
    :yellow [☶ Description]
    A regular container (27 slots).
    """

    material = BlockMaterials.SOLID

    base = BlockData.Base.CONTAINER
    sound = BlockData.Sound.WOOD
    facing = BlockData.Facing.PLAYER
    recipe = [
        BlockData.RecipeEntry("block", "[base]", 1),
    ]
    tags = [
        BlockData.Uses.BRIGHTNESS_FIX,
    ]


class Fridge(Block):
    """
    :yellow [☶ Description]
    A regular container (27 slots).

    :light_purple [⇄ Connectivity]
    Can connect to other fridges placed on top.
    """

    base = BlockData.Base.CONTAINER
    sound = BlockData.Sound.INDUSTRIAL
    facing = BlockData.Facing.PLAYER
    recipe = [
        BlockData.RecipeEntry("block", "iron_block", 4),
    ]
    tags = [
        BlockData.Uses.BLOCKSTATES,
        BlockData.Uses.BRIGHTNESS_FIX,
    ]
    blockstates = BlockData.BlockStates(
        "@self",
        BlockData.State("default"),
        BlockData.State("freezer", "~~~~1~"),
    )


class Washer(Block):
    """
    :yellow [☶ Description]
    Place a block of water on top to refill. Cleans color from leather armor.
    """

    base = BlockData.Base.HOPPER
    sound = BlockData.Sound.INDUSTRIAL
    facing = BlockData.Facing.PLAYER
    recipe = [
        BlockData.RecipeEntry("block", "iron_block", 4),
        BlockData.RecipeEntry("item", "bucket"),
    ]
    tags = [
        BlockData.Uses.TICK,
    ]
    blockstates = BlockData.BlockStates(
        "@self",
        BlockData.State("default"),
        BlockData.State("full"),
    )


class Trashcan(Block):
    """
    :yellow [☶ Description]
    Empty.
    """

    base = BlockData.Base.DRIPSTONE
    sound = BlockData.Sound.WOOD
    facing = BlockData.Facing.NONE
    recipe = [
        BlockData.RecipeEntry("block", "iron_block", 4),
    ]
    tags = [
        BlockData.Uses.PLACE,
        BlockData.Uses.BRIGHTNESS_FIX,
    ]


class TableLamp(BlockVariant):
    """
    :yellow [☶ Description]
    A light source.
    """

    material = BlockMaterials.WOOL
    block_type = BlockType.light(light_level=15)

    base = BlockData.Base.NONE
    sound = BlockData.Sound.WOOD
    facing = BlockData.Facing.NONE
    recipe = [
        BlockData.RecipeEntry("block", "[color]", 4),
    ]
    tags = [
        BlockData.Uses.PLACE,
        BlockData.Uses.NO_BASE,
    ]


class StreetLight(BlockVariant):
    """
    :yellow [☶ Description]
    A light source.
    """

    material = BlockMaterials.SOLID
    block_type = BlockType.light(light_level=15)

    base = BlockData.Base.NONE
    sound = BlockData.Sound.WOOD
    facing = BlockData.Facing.NORMAL
    recipe = [
        BlockData.RecipeEntry("block", "[base]", 1),
    ]
    tags = [
        BlockData.Uses.PLACE,
        BlockData.Uses.BLOCKSTATES,
        BlockData.Uses.NO_BASE,
    ]
    blockstates = BlockData.BlockStates(
        "!#no_solid_collision",
        BlockData.State("default"),
        BlockData.State("wall", "~1~~~~"),
    )


class DeskLamp(BlockVariant):
    """
    :yellow [☶ Description]
    A light source.
    """

    material = BlockMaterials.WOOL
    block_type = BlockType.light(light_level=15)

    base = BlockData.Base.NONE
    sound = BlockData.Sound.WOOD
    facing = BlockData.Facing.PLAYER
    recipe = [
        BlockData.RecipeEntry("block", "[color]", 4),
        BlockData.RecipeEntry("item", "glowstone_dust"),
    ]
    tags = [
        BlockData.Uses.PLACE,
        BlockData.Uses.NO_BASE,
    ]


class UpholsteryCouch(BlockVariant):
    """
    :yellow [☶ Description]
    A seat.

    :light_purple [⇄ Connectivity]
    Can connect to other couches to the left/right of it.
    """

    material = BlockMaterials.WOOD_WITH_WOOL
    block_type = BlockType.seat(width=0.9, height=0.5)

    base = BlockData.Base.TRIPWIRE
    sound = BlockData.Sound.WOOD
    facing = BlockData.Facing.PLAYER
    recipe = [
        BlockData.RecipeEntry("block", "[base]", 1),
        BlockData.RecipeEntry("block", "[color]", 6),
    ]
    tags = [
        BlockData.Uses.PLACE,
        BlockData.Uses.BLOCKSTATES,
    ]
    blockstates = BlockData.BlockStates(
        "@self",
        BlockData.State("default"),
        BlockData.State("left", "~~01~~"),
        BlockData.State("right", "~~10~~"),
        BlockData.State("middle", "~~11~~"),
    )


class FloorLamp(BlockVariant):
    """
    :yellow [☶ Description]
    A light source.
    """

    material = BlockMaterials.WOOL
    block_type = BlockType.light(light_level=15)

    base = BlockData.Base.VOID
    sound = BlockData.Sound.WOOL
    facing = BlockData.Facing.NONE
    recipe = [
        BlockData.RecipeEntry("item", "glowstone_dust"),
        BlockData.RecipeEntry("block", "[color]", 4),
    ]
    tags = [
        BlockData.Uses.PLACE,
        BlockData.Uses.NO_BASE,
    ]


class Curtains(Block):
    """
    :yellow [☶ Description]
    Sneak+Click to open/close.

    :light_purple [⇄ Connectivity]
    Can connect to other curtains.
    """

    base = BlockData.Base.VOID
    sound = BlockData.Sound.WOOL
    facing = BlockData.Facing.WALL_NORMAL
    recipe = [
        BlockData.RecipeEntry("block", "red_wool", 6),
    ]
    tags = [
        BlockData.Uses.PLACE,
        BlockData.Uses.NO_BASE,
        BlockData.Uses.CUSTOM_BASE,
    ]
    blockstates = BlockData.BlockStates(
        "@self",
        BlockData.State("top_closed"),
        BlockData.State("top_left"),
        BlockData.State("top_right"),
        BlockData.State("bottom_closed"),
        BlockData.State("bottom_left"),
        BlockData.State("bottom_right"),
        BlockData.State("middle"),
        BlockData.State("empty"),
    )


class Sink(BlockVariant):
    """
    :yellow [☶ Description]
    Click to toggle on/off.
    """

    material = BlockMaterials.SOLID

    base = BlockData.Base.UPPER
    sound = BlockData.Sound.WOOD
    facing = BlockData.Facing.WALL_NORMAL
    recipe = [
        BlockData.RecipeEntry("block", "[base]", 1),
    ]
    tags = [
        BlockData.Uses.PLACE,
        BlockData.Uses.TICK,
    ]
    blockstates = BlockData.BlockStates(
        "<manual>",
        BlockData.State("default"),
        BlockData.State("water"),
    )


class Toilet(BlockVariant):
    """
    :yellow [☶ Description]
    A seat.
    """

    material = BlockMaterials.SOLID
    block_type = BlockType.seat(width=0.6, height=0.51)

    base = BlockData.Base.TRIPWIRE
    sound = BlockData.Sound.WOOD
    facing = BlockData.Facing.PLAYER
    recipe = [
        BlockData.RecipeEntry("block", "[base]", 1),
    ]
    tags = [BlockData.Uses.PLACE]

File: bubblellaneous/furniture.go
package bubblellaneous

import (
	"github.com/bbfh-dev/bubblellaneous-pack/bubblellaneous/unit"
	"github.com/bbfh-dev/bubblellaneous-pack/lib/field"
)

var Furniture = unit.NewCategory().
	Add(unit.NewBlock(
		"bubble_bench",
		field.BASE_CONTAINER,
		field.SOUND_INDUSTRIAL,
		field.FACING_PLAYER,
		[]field.RecipeEntry{
			field.NewRecipeEntry("block", "crafting_table", 1),
			field.NewRecipeEntry("item", "glass_bottle", 1),
		},
		field.USE_GUI,
		field.USE_BRIGHTNESS_FIX,
	)).
	Add(unit.NewVariant(func() unit.Unit {
		return unit.NewBlock(
			"table",
			field.BASE_UPPER,
			field.SOUND_WOOD,
			field.FACING_NONE,
			[]field.RecipeEntry{
				field.NewRecipeEntry("block", "[base]", 1),
			},
			field.USE_PLACE,
			field.USE_BLOCKSTATES,
		).WithBlockstates(
			"@self", field.NewBlockState("default"),
			field.NewBlockState("side", "0100~~:0", "1000~~:180", "0010~~:-90", "0001~~:90"),
			field.NewBlockState("corner", "1010~~:0", "1001~~:-90", "0110~~:90", "0101~~:180"),
			field.NewBlockState("middle", "11~~~~", "~~11~~"),
		)
	}, field.SOLID_MATERIAL).Units()...).
	Add(unit.NewVariant(func() unit.Unit {
		return unit.NewBlock(
			"shelf",
			field.BASE_VOID,
			field.SOUND_WOOD,
			field.FACING_WALL_NORMAL,
			[]field.RecipeEntry{
				field.NewRecipeEntry("block", "[base]", 1),
			},
			field.USE_PLACE,
		).WithBlockType(field.ShelfBlock(1, 0))
	}, field.SOLID_MATERIAL).Units()...).
	Add(unit.NewVariant(func() unit.Unit {
		return unit.NewBlock(
			"cabinet",
			field.BASE_CONTAINER,
			field.SOUND_WOOD,
			field.FACING_PLAYER,
			[]field.RecipeEntry{
				field.NewRecipeEntry("block", "[base]", 1),
			},
			field.USE_BRIGHTNESS_FIX,
		)
	}, field.SOLID_MATERIAL).Units()...).
	Add(unit.NewVariant(func() unit.Unit {
		return unit.NewBlock(
			"drawer",
			field.BASE_UPPER,
			field.SOUND_WOOD,
			field.FACING_PLAYER,
			[]field.RecipeEntry{
				field.NewRecipeEntry("block", "[base]", 1),
			},
		)
	}, field.SOLID_MATERIAL).Units()...).
	Add(unit.NewVariant(func() unit.Unit {
		return unit.NewBlock(
			"cupboard",
			field.BASE_SOLID,
			field.SOUND_WOOD,
			field.FACING_PLAYER,
			[]field.RecipeEntry{
				field.NewRecipeEntry("block", "[base]", 1),
			},
			field.USE_PLACE,
			field.USE_BRIGHTNESS_FIX,
		).WithBlockType(field.ShelfBlock(2, 0))
	}, field.SOLID_MATERIAL).Units()...).
	Add(unit.NewVariant(func() unit.Unit {
		return unit.NewBlock(
			"bench",
			field.BASE_TRIPWIRE,
			field.SOUND_WOOD,
			field.FACING_PLAYER,
			[]field.RecipeEntry{
				field.NewRecipeEntry("block", "[base]", 1),
			},
			field.USE_PLACE,
		).WithBlockType(field.SeatBlock(0.8, 0.6))
	}, field.SOLID_MATERIAL).Units()...).
	Add(unit.NewVariant(func() unit.Unit {
		return unit.NewBlock(
			"couch",
			field.BASE_TRIPWIRE,
			field.SOUND_WOOL,
			field.FACING_PLAYER,
			[]field.RecipeEntry{
				field.NewRecipeEntry("block", "[color]", 6),
			},
			field.USE_PLACE,
			field.USE_BLOCKSTATES,
		).WithBlockstates(
			"@self", field.NewBlockState("default"),
			field.NewBlockState("left", "0010~~"),
			field.NewBlockState("right", "0001~~"),
			field.NewBlockState("middle", "0011~~"),
			field.NewBlockState("corner", "1010~~:90", "1001~~:0"),
			field.NewBlockState("slab", "~1~~~~"),
		).WithBlockType(field.SeatBlock(0.8, 0.5))
	}, field.WOOL_MATERIAL).Units()...).
	Add(unit.NewVariant(func() unit.Unit {
		return unit.NewBlock(
			"stool",
			field.BASE_VOID,
			field.SOUND_WOOD,
			field.FACING_NONE,
			[]field.RecipeEntry{
				field.NewRecipeEntry("block", "[base]", 1),
			},
			field.USE_PLACE,
		).WithBlockType(field.SeatBlock(0.8, 0.6))
	}, field.WOOD_MATERIAL).Units()...).
	Add(unit.NewVariant(func() unit.Unit {
		return unit.NewBlock(
			"barstool",
			field.BASE_VOID,
			field.SOUND_WOOD,
			field.FACING_NONE,
			[]field.RecipeEntry{
				field.NewRecipeEntry("block", "[base]", 1),
			},
			field.USE_PLACE,
		).WithBlockType(field.SeatBlock(0.8, 0.7))
	}, field.WOOD_MATERIAL).Units()...).
	Add(unit.NewVariant(func() unit.Unit {
		return unit.NewBlock(
			"dinning_chair",
			field.BASE_VOID,
			field.SOUND_WOOD,
			field.FACING_PLAYER,
			[]field.RecipeEntry{
				field.NewRecipeEntry("block", "[base]", 1),
				field.NewRecipeEntry("block", "[color]", 2),
			},
			field.USE_PLACE,
		).WithBlockType(field.SeatBlock(0.8, 0.46))
	}, field.WOOD_WITH_WOOL_MATERIAL()).Units()...).
	Add(unit.NewVariant(func() unit.Unit {
		return unit.NewBlock(
			"blinds",
			field.BASE_VOID,
			field.SOUND_WOOD,
			field.FACING_WALL_NORMAL,
			[]field.RecipeEntry{
				field.NewRecipeEntry("block", "[base]", 1),
			},
		)
	}, field.SOLID_MATERIAL).Units()...).
	Add(unit.NewVariant(func() unit.Unit {
		return unit.NewBlock(
			"kitchen_cabinet",
			field.BASE_CONTAINER,
			field.SOUND_WOOD,
			field.FACING_PLAYER,
			[]field.RecipeEntry{
				field.NewRecipeEntry("block", "[base]", 1),
			},
			field.USE_BRIGHTNESS_FIX,
		)
	}, field.SOLID_MATERIAL).Units()...).
	Add(unit.NewBlock(
		"fridge",
		field.BASE_CONTAINER,
		field.SOUND_INDUSTRIAL,
		field.FACING_PLAYER,
		[]field.RecipeEntry{
			field.NewRecipeEntry("block", "iron_block", 4),
		},
		field.USE_BLOCKSTATES,
		field.USE_BRIGHTNESS_FIX,
	).WithBlockstates(
		"@self", field.NewBlockState("default"),
		field.NewBlockState("freezer", "~~~~1~"),
	)).
	Add(unit.NewBlock(
		"washer",
		field.BASE_HOPPER,
		field.SOUND_INDUSTRIAL,
		field.FACING_PLAYER,
		[]field.RecipeEntry{
			field.NewRecipeEntry("block", "iron_block", 4),
			field.NewRecipeEntry("item", "bucket", 1),
		},
		field.USE_TICK,
	).WithBlockstates(
		"@self", field.NewBlockState("default"),
		field.NewBlockState("full"),
	)).
	Add(unit.NewBlock(
		"trashcan",
		field.BASE_DRIPSTONE,
		field.SOUND_WOOD,
		field.FACING_NONE,
		[]field.RecipeEntry{
			field.NewRecipeEntry("block", "iron_block", 4),
		},
		field.USE_PLACE,
		field.USE_BRIGHTNESS_FIX,
	)).
	Add(unit.NewVariant(func() unit.Unit {
		return unit.NewBlock(
			"table_lamp",
			field.BASE_NONE,
			field.SOUND_WOOD,
			field.FACING_NONE,
			[]field.RecipeEntry{
				field.NewRecipeEntry("block", "[color]", 4),
			},
			field.USE_PLACE,
			field.USE_NO_BASE,
		).WithBlockType(field.LightBlock(15))
	}, field.WOOL_MATERIAL).Units()...).
	Add(unit.NewVariant(func() unit.Unit {
		return unit.NewBlock(
			"street_light",
			field.BASE_NONE,
			field.SOUND_WOOD,
			field.FACING_NORMAL,
			[]field.RecipeEntry{
				field.NewRecipeEntry("block", "[base]", 1),
			},
			field.USE_PLACE,
			field.USE_BLOCKSTATES,
			field.USE_NO_BASE,
		).WithBlockstates(
			"!#no_solid_collision",
			field.NewBlockState("default"),
			field.NewBlockState("wall", "~1~~~~"),
		).WithBlockType(field.LightBlock(15))
	}, field.SOLID_MATERIAL).Units()...).
	Add(unit.NewVariant(func() unit.Unit {
		return unit.NewBlock(
			"desk_lamp",
			field.BASE_NONE,
			field.SOUND_WOOD,
			field.FACING_PLAYER,
			[]field.RecipeEntry{
				field.NewRecipeEntry("block", "[color]", 4),
				field.NewRecipeEntry("item", "glowstone_dust", 1),
			},
			field.USE_PLACE,
			field.USE_NO_BASE,
		).WithBlockType(field.LightBlock(15))
	}, field.WOOL_MATERIAL).Units()...).
	Add(unit.NewVariant(func() unit.Unit {
		return unit.NewBlock(
			"upholstery_couch",
			field.BASE_TRIPWIRE,
			field.SOUND_WOOD,
			field.FACING_PLAYER,
			[]field.RecipeEntry{
				field.NewRecipeEntry("block", "[base]", 1),
				field.NewRecipeEntry("block", "[color]", 6),
			},
			field.USE_PLACE,
			field.USE_BLOCKSTATES,
		).WithBlockstates(
			"@self", field.NewBlockState("default"),
			field.NewBlockState("left", "~~01~~"),
			field.NewBlockState("right", "~~10~~"),
			field.NewBlockState("middle", "~~11~~"),
		).WithBlockType(field.SeatBlock(0.9, 0.5))
	}, field.WOOD_WITH_WOOL_MATERIAL()).Units()...).
	Add(unit.NewVariant(func() unit.Unit {
		return unit.NewBlock(
			"floor_lamp",
			field.BASE_VOID,
			field.SOUND_WOOL,
			field.FACING_NONE,
			[]field.RecipeEntry{
				field.NewRecipeEntry("item", "glowstone_dust", 1),
				field.NewRecipeEntry("block", "[color]", 4),
			},
			field.USE_PLACE,
			field.USE_NO_BASE,
		).WithBlockType(field.LightBlock(15))
	}, field.WOOL_MATERIAL).Units()...).
	Add(unit.NewBlock(
		"curtains",
		field.BASE_VOID,
		field.SOUND_WOOL,
		field.FACING_WALL_NORMAL,
		[]field.RecipeEntry{
			field.NewRecipeEntry("block", "red_wool", 6),
		},
		field.USE_PLACE,
		field.USE_NO_BASE,
		field.USE_CUSTOM_BASE,
	).WithBlockstates(
		"@self", field.NewBlockState("top_closed"),
		field.NewBlockState("top_left"),
		field.NewBlockState("top_right"),
		field.NewBlockState("bottom_closed"),
		field.NewBlockState("bottom_left"),
		field.NewBlockState("bottom_right"),
		field.NewBlockState("middle"),
		field.NewBlockState("empty"),
	)).
	Add(unit.NewVariant(func() unit.Unit {
		return unit.NewBlock(
			"sink",
			field.BASE_UPPER,
			field.SOUND_WOOD,
			field.FACING_WALL_NORMAL,
			[]field.RecipeEntry{
				field.NewRecipeEntry("block", "[base]", 1),
			},
			field.USE_PLACE,
			field.USE_TICK,
		).WithBlockstates(
			"<manual>", field.NewBlockState("default"),
			field.NewBlockState("water"),
		)
	}, field.SOLID_MATERIAL).Units()...).
	Add(unit.NewVariant(func() unit.Unit {
		return unit.NewBlock(
			"toilet",
			field.BASE_TRIPWIRE,
			field.SOUND_WOOD,
			field.FACING_PLAYER,
			[]field.RecipeEntry{
				field.NewRecipeEntry("block", "[base]", 1),
			},
			field.USE_PLACE,
		).WithBlockType(field.SeatBlock(0.6, 0.51))
	}, field.SOLID_MATERIAL).Units()...)
